File: versioncontrol/src/main/java/ir/ayantech/versioncontrol/VersionControlDialog.java
package ir.ayantech.versioncontrol;

import android.app.Activity;
import android.app.Dialog;
import android.content.Context;
import android.content.Intent;
import android.graphics.Typeface;
import android.net.Uri;
import android.os.Build;
import android.os.Environment;
import androidx.annotation.NonNull;
import androidx.core.content.FileProvider;
import android.util.Log;
import android.view.View;
import android.view.Window;
import android.widget.LinearLayout;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.coolerfall.download.DownloadCallback;
import com.coolerfall.download.DownloadManager;
import com.coolerfall.download.DownloadRequest;
import com.coolerfall.download.OkHttpDownloader;
import com.coolerfall.download.Priority;

import java.io.File;
import java.util.ArrayList;
import java.util.Date;
import java.util.concurrent.TimeUnit;

import ir.ayantech.versioncontrol.api.CheckVersion;
import ir.ayantech.versioncontrol.api.GetLastVersion;

public class VersionControlDialog extends Dialog {

    private int id = -1;
    private DownloadManager manager;

    public VersionControlDialog(@NonNull final Activity activity,
                                String title,
                                String message,
                                String positiveButton,
                                String negativeButton,
                                ArrayList<String> changeLogs,
                                final String linkType,
                                final String link,
                                final String updateStatus,
                                final Typeface typeface) {
        super(activity);
        requestWindowFeature(Window.FEATURE_NO_TITLE);
        setContentView(R.layout.activity_version_control);
        this.setCancelable(false);
        this.setCanceledOnTouchOutside(false);

        ((TextView) findViewById(R.id.titleTv)).setText(title);
        ((TextView) findViewById(R.id.messageTv)).setText(message);
        ((TextView) findViewById(R.id.positiveTv)).setText(positiveButton);
        ((TextView) findViewById(R.id.negativeTv)).setText(negativeButton);
        if (changeLogs == null)
            findViewById(R.id.changeLogTv).setVisibility(View.GONE);
        else if (changeLogs.isEmpty())
            findViewById(R.id.changeLogTv).setVisibility(View.GONE);
        else {
            StringBuilder changeLog = new StringBuilder();
            for (String s : changeLogs) {
                changeLog.append(s).append("\n");
            }
            ((TextView) findViewById(R.id.changeLogTv)).setText(changeLog);
        }
        manager = new DownloadManager.Builder().context(getContext())
                .downloader(OkHttpDownloader.create())
                .threadPoolSize(2)
                .build();
        findViewById(R.id.positiveTv).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    if (GetLastVersion.LinkType.DIRECT.contentEquals(linkType)) {
                        if (getRootDirPath(getContext()) == null) {
                            openUrl(getContext(), link);
                            return;
                        }
                        findViewById(R.id.progressBar).setVisibility(View.VISIBLE);
                        findViewById(R.id.progressTv).setVisibility(View.VISIBLE);
                        String destPath = getRootDirPath(getContext()) + "/newversion" + String.valueOf(new Date().getTime()) + ".apk";
                        DownloadRequest request = new DownloadRequest.Builder()
                                .url(link)
                                .retryTime(5)
                                .retryInterval(2, TimeUnit.SECONDS)
                                .progressInterval(100, TimeUnit.MILLISECONDS)
                                .priority(Priority.HIGH)
                                .destinationFilePath(destPath)
                                .downloadCallback(new DownloadCallback() {
                                    @Override
                                    public void onStart(int downloadId, long totalBytes) {
                                    }

                                    @Override
                                    public void onRetry(int downloadId) {
                                    }

                                    @Override
                                    public void onProgress(int downloadId, long bytesWritten, long totalBytes) {
                                        long progressPercent = bytesWritten * 100 / totalBytes;
                                        ((ProgressBar) findViewById(R.id.progressBar)).setProgress((int) progressPercent);
                                        ((TextView) findViewById(R.id.progressTv)).setText(String.format("%%%s", String.valueOf(progressPercent)));
                                    }

                                    @Override
                                    public void onSuccess(int downloadId, String filePath) {
                                        try {
                                            installApp(getContext(), filePath);
                                            dismiss();
                                            if (CheckVersion.UpdateStatus.MANDATORY.contentEquals(updateStatus)) {
                                                activity.finish();
                                            }
                                        } catch (Exception e) {
                                            e.printStackTrace();
                                        }
                                    }

                                    @Override
                                    public void onFailure(int downloadId, int statusCode, String errMsg) {
                                        Log.e("AyanVC:", errMsg);
                                        dismiss();
                                        openUrl(getContext(), link);
                                    }
                                })
                                .build();

                        id = manager.add(request);
                    } else if (GetLastVersion.LinkType.PAGE.contentEquals(linkType)) {
                        dismiss();
                        openUrl(getContext(), link);
                        if (CheckVersion.UpdateStatus.MANDATORY.contentEquals(updateStatus))
                            activity.finish();
                    }
                } catch (Exception e) {
                    dismiss();
                    openUrl(getContext(), link);
                }
            }
        });
        findViewById(R.id.negativeTv).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                try {
                    manager.cancel(id);
                } catch (Exception e) {
                }
                dismiss();
                if (CheckVersion.UpdateStatus.MANDATORY.contentEquals(updateStatus))
                    activity.finish();
            }
        });

        if (typeface != null) {
            LinearLayout mainContentLl = findViewById(R.id.mainContentLl);
            for (int i = 0; i < mainContentLl.getChildCount(); i++) {
                if (mainContentLl.getChildAt(i) instanceof TextView) {
                    ((TextView) mainContentLl.getChildAt(i)).setTypeface(typeface);
                }
            }
            ((TextView) findViewById(R.id.positiveTv)).setTypeface(typeface);
            ((TextView) findViewById(R.id.negativeTv)).setTypeface(typeface);
        }
    }

    @Override
    public void onBackPressed() {
    }

    private void installApp(Context context, String path) {
        File toInstall = new File(path);

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            try {
                Uri apkUri = FileProvider.getUriForFile(context, context.getPackageName() + ".provider", toInstall);
                Intent intent = new Intent(Intent.ACTION_INSTALL_PACKAGE);
                intent.setData(apkUri);
                intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
                context.startActivity(intent);
            } catch (Exception e1) {
                e1.printStackTrace();
            }
        } else {
            Uri apkUri = Uri.fromFile(toInstall);
            Intent intent = new Intent(Intent.ACTION_VIEW);
            intent.setDataAndType(apkUri, "application/vnd.android.package-archive");
            intent.setFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            intent.setFlags(Intent.FLAG_GRANT_READ_URI_PERMISSION);
            context.startActivity(intent);
        }
    }

    private void openUrl(Context context, String url) {
        Intent browserIntent = new Intent(Intent.ACTION_VIEW, Uri.parse(url));
        context.startActivity(browserIntent);
    }

    private String getRootDirPath(Context context) {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N)
            return context.getFilesDir().getAbsolutePath();
        else if (Environment.MEDIA_MOUNTED.equals(Environment.getExternalStorageState()))
            return context.getExternalFilesDir(null).getAbsolutePath();
        else
            return null;
    }
}

File: versioncontrol/src/main/java/ir/ayantech/versioncontrol/VersionControlDialog.kt
package ir.ayantech.versioncontrol

import android.app.Activity
import android.app.Dialog
import android.content.Context
import android.content.Intent
import android.graphics.Typeface
import android.net.Uri
import android.os.Build
import android.os.Environment
import android.view.View
import android.view.Window
import android.widget.LinearLayout
import android.widget.ProgressBar
import android.widget.TextView
import androidx.core.content.FileProvider
import androidx.core.net.toUri
import com.coolerfall.download.DownloadCallback
import com.coolerfall.download.DownloadManager
import com.coolerfall.download.DownloadRequest
import com.coolerfall.download.OkHttpDownloader
import com.coolerfall.download.Priority
import ir.ayantech.versioncontrol.api.CheckVersion
import ir.ayantech.versioncontrol.api.GetLastVersion
import java.io.File
import java.util.Date
import java.util.concurrent.TimeUnit

class VersionControlDialog(
    private val activity: Activity,
    title: String?,
    message: String?,
    positiveButton: String?,
    negativeButton: String?,
    changeLogs: ArrayList<String>?,
    linkType: String?,
    link: String?,
    updateStatus: String?,
    typeface: Typeface?
) : Dialog(activity) {

    private var id = -1
    private var manager: DownloadManager

    init {
        requestWindowFeature(Window.FEATURE_NO_TITLE)
        setContentView(R.layout.activity_version_control)
        setCancelable(false)
        setCanceledOnTouchOutside(false)

        findViewById<TextView>(R.id.titleTv).text = title
        findViewById<TextView>(R.id.messageTv).text = message
        findViewById<TextView>(R.id.positiveTv).text = positiveButton
        findViewById<TextView>(R.id.negativeTv).text = negativeButton

        if (changeLogs.isNullOrEmpty()) {
            findViewById<View>(R.id.changeLogTv).visibility = View.GONE
        } else {
            val changeLog = StringBuilder()
            for (s in changeLogs) {
                changeLog.append(s).append("\n")
            }
            findViewById<TextView>(R.id.changeLogTv).text = changeLog
        }

        manager = DownloadManager.Builder()
            .context(context)
            .downloader(OkHttpDownloader.create())
            .threadPoolSize(2)
            .build()

        findViewById<View>(R.id.positiveTv).setOnClickListener {
            try {
                if (GetLastVersion.LinkType.DIRECT == linkType) {
                    val rootDirPath = getRootDirPath(context)
                    if (rootDirPath == null) {
                        openUrl(context, link)
                        return@setOnClickListener
                    }
                    findViewById<View>(R.id.progressBar).visibility = View.VISIBLE
                    findViewById<View>(R.id.progressTv).visibility = View.VISIBLE
                    val destPath = "$rootDirPath/newversion${Date().time}.apk"
                    val request = DownloadRequest.Builder()
                        .url(link)
                        .retryTime(5)
                        .retryInterval(2, TimeUnit.SECONDS)
                        .progressInterval(100, TimeUnit.MILLISECONDS)
                        .priority(Priority.HIGH)
                        .destinationFilePath(destPath)
                        .downloadCallback(object : DownloadCallback() {
                            override fun onStart(downloadId: Int, totalBytes: Long) {}
                            override fun onRetry(downloadId: Int) {}

                            override fun onProgress(downloadId: Int, bytesWritten: Long, totalBytes: Long) {
                                val progressPercent = (bytesWritten * 100) / totalBytes
                                findViewById<ProgressBar>(R.id.progressBar).progress = progressPercent.toInt()
                                findViewById<TextView>(R.id.progressTv).text = String.format("%%%s", progressPercent)
                            }

                            override fun onSuccess(downloadId: Int, filePath: String) {
                                try {
                                    installApp(context, filePath)
                                    dismiss()
                                    if (CheckVersion.UpdateStatus.MANDATORY == updateStatus) {
                                        activity.finish()
                                    }
                                } catch (e: Exception) {
                                    e.printStackTrace()
                                }
                            }

                            override fun onFailure(downloadId: Int, statusCode: Int, errMsg: String?) {
                                dismiss()
                                openUrl(context, link)
                            }
                        })
                        .build()

                    id = manager.add(request)
                } else if (GetLastVersion.LinkType.PAGE == linkType) {
                    dismiss()
                    openUrl(context, link)
                    if (CheckVersion.UpdateStatus.MANDATORY == updateStatus) {
                        activity.finish()
                    }
                }
            } catch (_: Exception) {
                dismiss()
                openUrl(context, link)
            }
        }

        findViewById<View>(R.id.negativeTv).setOnClickListener {
            try {
                manager.cancel(id)
            } catch (_: Exception) {
            }
            dismiss()
            if (CheckVersion.UpdateStatus.MANDATORY == updateStatus) {
                activity.finish()
            }
        }

        if (typeface != null) {
            val mainContentLl = findViewById<LinearLayout>(R.id.mainContentLl)
            if (mainContentLl != null) {
                for (i in 0 until mainContentLl.childCount) {
                    val child = mainContentLl.getChildAt(i)
                    if (child is TextView) {
                        child.typeface = typeface
                    }
                }
            }
            findViewById<TextView>(R.id.positiveTv).typeface = typeface
            findViewById<TextView>(R.id.negativeTv).typeface = typeface
        }
    }

    @Deprecated("Deprecated in Java / Android")
    override fun onBackPressed() {
    }

    private fun installApp(context: Context, path: String) {
        val toInstall = File(path)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            try {
                val apkUri = FileProvider.getUriForFile(context, "${context.packageName}.provider", toInstall)
                val intent = Intent(Intent.ACTION_VIEW).apply {
                    setDataAndType(apkUri, "application/vnd.android.package-archive")
                    flags = Intent.FLAG_GRANT_READ_URI_PERMISSION
                }
                context.startActivity(intent)
            } catch (e: Exception) {
                e.printStackTrace()
            }
        } else {
            val apkUri = Uri.fromFile(toInstall)
            val intent = Intent(Intent.ACTION_VIEW).apply {
                setDataAndType(apkUri, "application/vnd.android.package-archive")
                flags = Intent.FLAG_ACTIVITY_NEW_TASK or Intent.FLAG_GRANT_READ_URI_PERMISSION
            }
            context.startActivity(intent)
        }
    }

    private fun openUrl(context: Context, url: String?) {
        if (url.isNullOrEmpty()) return
        val browserIntent = Intent(Intent.ACTION_VIEW, url.toUri())
        context.startActivity(browserIntent)
    }

    private fun getRootDirPath(context: Context): String? {
        return if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            context.filesDir.absolutePath
        } else if (Environment.MEDIA_MOUNTED == Environment.getExternalStorageState()) {
            context.getExternalFilesDir(null)?.absolutePath
        } else {
            null
        }
    }
}
